README.md
<h1 align="center" style="border-bottom: none;">:chart_with_downwards_trend:&nbsp;&nbsp; MQTT ⇒ InfluxDB Gateway &nbsp;&nbsp;:chart_with_upwards_trend:</h1>
<h3 align="center">An MQTT gateway sending data to InfluxDB</h3>

<p align="center">
    <img alt="CircleCI" src="https://img.shields.io/circleci/build/github/radeklat/mqtt-influxdb-gateway">
    <img alt="Codecov" src="https://img.shields.io/codecov/c/github/radeklat/mqtt-influxdb-gateway">
    <img alt="GitHub tag (latest SemVer)" src="https://img.shields.io/github/tag/radeklat/mqtt-influxdb-gateway">
    <img alt="Maintenance" src="https://img.shields.io/maintenance/yes/2023">
    <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/radeklat/mqtt-influxdb-gateway">
    <img alt="Docker Pulls" src="https://img.shields.io/docker/pulls/radeklat/mqtt-influxdb-gateway">
    <img alt="Docker Image Version (latest semver)" src="https://img.shields.io/docker/v/radeklat/mqtt-influxdb-gateway?label=image%20version">
    <img alt="Docker Image Size (latest semver)" src="https://img.shields.io/docker/image-size/radeklat/mqtt-influxdb-gateway">
</p>

# Installation

## As a script

1. Open terminal
2. Make sure [`git`](https://github.com/git-guides/install-git) is installed:
   ```shell script
   git --version   
   ```
   The output should look something like:
   ```text
   git version 2.25.1
   ```
3. Make sure Python 3.9+ is installed:
   ```shell script
   python --version
   ```
   The output should look something like:
   ```text
   Python 3.9.7
   ```
   * If it shows `2.7.x` instead, try `python3` instead and use it in the rest of the guide.
   * If it shows `3.8.x` or lower, use [`pyenv`](https://github.com/pyenv/pyenv#installation) to install a higher version of Python on your system.
4. Make sure [`poetry`](https://python-poetry.org/docs/#installation) is installed:
   ```shell script
   poetry --version
   ```
   The output should look something like:
   ```
   Poetry version 1.1.5
5. Clone this repository:
   ```shell script
   git clone https://github.com/radeklat/mqtt-influxdb-gateway.git
   cd mqtt-influxdb-gateway
   ```
6. Copy the `.env.template` file into `.env` file:
   ```shell script
   cp .env.template .env
   ```
7. Edit the `.env` file, fill all missing values and/or change existing ones to suit your needs.
8. Install all application dependencies:
   ```shell script
   poetry install --no-dev
   ```
9. Run the app:
    ```shell script
    poetry run python src/main.py
    ```

## As a docker container

1. Open terminal
2. Make sure [`docker`](https://docs.docker.com/get-docker/) is installed:
   ```shell script
   docker --version   
   ```
   The output should look something like:
   ```text
   Docker version 20.10.5, build 55c4c88
   ```
3. Create a base folder structure for the service:
   ```shell script
   cd mqtt-influxdb-gateway
   # Download .env file template
   curl https://raw.githubusercontent.com/radeklat/mqtt-influxdb-gateway/main/.env.template --output .env
   ```
4. Edit the `.env` file and fill the missing details.
5. Run the container:
   ```shell script
   docker run \
      --pull always --rm --name mqtt-influxdb-gateway \
      --env-file=.env \
      radeklat/mqtt-influxdb-gateway:latest
   ```
   This configuration will run the service in the foreground (you need to keep the terminal open) and always use the latest version.
   * Change `latest` to [a specific version](https://hub.docker.com/repository/registry-1.docker.io/radeklat/mqtt-influxdb-gateway/tags) if you don't always want the latest version or remove the `--pull always` flag to not update.


Add `--detach` flag to run in the background. You can close the terminal, but it will not start on system start up.
* To see the log of the service running in the background, run `docker logs mqtt-influxdb-gateway`
* To stop the service, run `docker container stop mqtt-influxdb-gateway`

## As a service

You can use the above-mentioned docker image to run the sync as a service on server or even your local machine. The simples way is to use docker compose:

1. Make sure you have [`docker`](https://docs.docker.com/get-docker/) and [`docker-compose`](https://docs.docker.com/compose/install/) installed:
   ```shell script
   docker --version
   docker-compose --version
   ```
   The output should look something like:
   ```text
   Docker version 20.10.5, build 55c4c88
   docker-compose version 1.28.5, build unknown
   docker-py version: 4.4.4
   CPython version: 3.8.5
   OpenSSL version: OpenSSL 1.1.1f  31 Mar 2020
   ```
2. Create a base folder structure for the service:
   ```shell script
   mkdir mqtt-influxdb-gateway 
   cd mqtt-influxdb-gateway
   ```
3. Download a compose file and an example .env file:
   ```shell script
   BASE_URL=https://raw.githubusercontent.com/radeklat/mqtt-influxdb-gateway/main
   curl $BASE_URL/.env.template --output .env
   curl $BASE_URL/docker-compose.yml -O
   ```
4. Edit the `.env` file and fill the missing details.
5. Run the service:
   ```shell script
   docker-compose up
   ```
   This command will run the service in the foreground (you need to keep the terminal open) and always use the latest version.
   * Change `latest` to [a specific version](https://hub.docker.com/repository/registry-1.docker.io/radeklat/mqtt-influxdb-gateway/tags) if you don't always want the latest version.


Add `--detach` flag to run in the background. You can close the terminal. The service should start on system start up.
* To see the log of the service running in the background, run `docker-compose logs mqtt-influxdb-gateway`
* To stop the service, run `docker-compose stop mqtt-influxdb-gateway`

# Update

## As a script

1. Open terminal
2. Navigate to the project repository:
   ```shell script
   cd mqtt-influxdb-gateway
   ```
3. Update the source code:
   ```shell script
   git pull
   ```

4. Update application dependencies:
   ```shell script
   poetry install --no-dev
   ```
5. Run the application again.

## From docker-compose

If you used the `latest` tag, run `docker-compose pull mqtt-influxdb-gateway`.

# Environment variables
<!--
    DO NOT EDIT the following section!
    It is auto-generated with settings-doc in CircleCI and committed
    back to the repository on changes.
-->
<!-- settings-doc start -->
## `LOG_LEVEL`

*Optional*, default value: `INFO`

Logging level.

### Possible values

`TRACE`, `DEBUG`, `INFO`, `SUCCESS`, `WARNING`, `ERROR`, `CRITICAL`

## `INFLUXDB_URL`

**Required**

URL of your InfluxDB Cloud instance.

## `INFLUXDB_ORGANIZATION_ID`

**Required**

Organization ID. A hexadecimal string. Should be part of the cloud instance URL or in About Organization. This is not the organization name!

## `INFLUXDB_API_TOKEN`

**Required**

API token with write access to `influxdb_default_bucket`.

## `INFLUXDB_PRECISION`

*Optional*, default value: `ns`

The precision for the unix timestamps within the body line-protocol.

### Possible values

`s`, `ms`, `ns`, `us`

## `INFLUXDB_DEFAULT_BUCKET`

*Optional*, default value: `None`

Default value if `{bucket}` in `mqtt_topic_pattern` below is not set or parsed.

## `INFLUXDB_DEFAULT_MEASUREMENT`

*Optional*, default value: `None`

Default value if `{measurement}` in `mqtt_topic_pattern` below is not set or parsed.

## `MQTT_HOST`

**Required**

IP address or a host name of an MQTT broker.

## `MQTT_PORT`

*Optional*, default value: `1883`

Port of the MQTT broker.

## `MQTT_USERNAME`

*Optional*, default value: `None`

User name used to authenticate with the MQTT broker.

## `MQTT_PASSWORD`

*Optional*, default value: `None`

Password used to authenticate with the MQTT broker.

## `MQTT_TOPIC_SUBSCRIBE`

**Required**

A topic to subscribe to. You can also use the '+' and '#' wildcards.

## `MQTT_TOPIC_PATTERN`

**Required**

A pattern to parse the topic into fields sent to InfluxDB. Use `{variable}` or `{variable_type:value}` syntax in topic parts to create the match. Use a plain string to ignore a topic part. The pattern must be a prefix of `mqtt_topic_subscribe`.

When both a variable and a `influxdb_*` configuration option are defined, variable takes precedence.

### Examples

`dt/influxdb/{bucket}/{measurement}/{tag:device_id}/{field}/{value_type}` will parse `dt/influxdb/home/environment/esp8266/temperature/float 23.412` into:

```json
{
   "bucket": "home",
   "measurement": "environment",
   "tags": {"device_id": "esp8266"},
   "fields": {"temperature": 23.412}
}
```

### Possible values

- `{bucket}`: InfluxDB bucket to send data to. Required unless `influxdb_default_bucket` is set.
- `{measurement}`: InfluxDB measurement field. Required unless `influxdb_default_measurement` is set.
- `{field}`: Field name. Required.
- `{value_type}`: Value type used to parse the received data. Optional.
  Acceptable values in the topic are `str`, `int`, `float` and `bool`. Default value is `str`.
- `{tag:TAG_NAME}`: where TAG_NAME will be used as a tag name and the parsed value as a value. Optional.

## `MQTT_MERGE_DATA_POINTS_ON`

*Optional*, default value: `{measurement}{bucket}{tags}`

It is expected there will be a single value in each topic. But InfluxDB allows multiple fields to be sent at once (for example from one device). Use this option to define criteria on which to merge data points. The order and any extra characters aside from variables below are irrelevant. All collected fields are then sent to InfluxDB when any of the fields is seen more than once (suggesting new data is coming in).

### Examples

A value of `{tags[device_id]}` will merge all received data that share the same `device_id` tag.

### Possible values

- `{bucket}`
- `{measurement}`
- `{tags}`: for all tags and their values
- `{tags[NAME]}`: for a value of single tag named `NAME`
<!-- settings-doc end -->

# Development

## Building for ARMv7

At the moment, build for ARMv7 takes about 15-20 minutes. This is caused by the cryptography package switching to Rust compiler (some info [here](https://github.com/matrix-org/synapse/issues/9403)). The build process is significantly slower and not suitable for running in the free tier CircleCI build pipeline.

To build and push the image manually, run:

```shell script
inv docker-build
```

## Paho documentation

https://github.com/eclipse/paho.mqtt.python#contents

## InfluxDB documentation

- https://docs.influxdata.com/influxdb/cloud/write-data/developer-tools/api/
- https://docs.influxdata.com/influxdb/cloud/api/#operation/PostWrite

# TODOs

- Define what to merge data points on.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
Types of changes are:

- **Breaking changes** for breaking changes.
- **Features** for new features or changes in existing functionality.
- **Fixes** for any bug fixes.
- **Deprecated** for soon-to-be removed features.

## [Unreleased]

## [0.2.4] - 2023-11-07

### Fixes

- Update `pydantic` to `2.x`.

## [0.2.3] - 2023-09-16

### Fixes

- Switch `settings-doc` from CI to `pre-commit` hook.
- 

## [0.2.2] - 2022-12-30

### Fixes

- Dependencies update

## [0.2.1] - 2021-11-07

### Fixes

- Improved environment variables documentation.

## [0.2.0] - 2021-11-02

### Features

- Option to define when to merge data points (`mqtt_merge_data_points_on`).
- Validation of variables in `mqtt_topic_pattern`.
- Validation of `influxdb_url`
- Options to set global bucket and measurement with `influxdb_default_bucket` and `influxdb_default_measurement`.
- Documentation of settings.

## [0.1.2] - 2021-10-31

### Features

- Change `Dockerfile` to use Python version from `pyproject.toml`.

## [0.1.1] - 2021-10-30

### Features

- Updates build process to build docker images for multiple architectures.

### Fixes

- docker-compose file not starting up the container correctly.

## [0.1.0] - 2021-10-28

- Initial release

[Unreleased]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/0.2.4...HEAD
[0.2.4]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/0.2.3...0.2.4
[0.2.3]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/0.2.2...0.2.3
[0.2.2]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/0.2.1...0.2.2
[0.2.1]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/0.1.2...0.2.0
[0.1.2]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/radeklat/mqtt-influxdb-gateway/compare/initial...0.1.0


tests/conftest.py
from pathlib import Path

import pytest
import toml
from delfino.constants import PYPROJECT_TOML_FILENAME
from delfino.models.pyproject_toml import PyprojectToml


@pytest.fixture(scope="session")
def project_root():
    return Path(__file__).parent.parent


@pytest.fixture(scope="session")
def pyproject_toml(project_root):
    return PyprojectToml(**toml.load(project_root / PYPROJECT_TOML_FILENAME))


@pytest.fixture(scope="session")
def poetry(pyproject_toml):
    assert pyproject_toml.tool.poetry
    return pyproject_toml.tool.poetry


src/main.py
import sys

from influxdb_client import InfluxDBClient
from influxdb_client.client.write_api import ASYNCHRONOUS
from loguru import logger
from paho.mqtt.client import Client as MQTTClient
from paho.mqtt.client import MQTTMessage
from pydantic import BaseModel, ConfigDict, Field

from constants import LOGGING_FORMAT
from influx_db import InfluxDBLine, MergeConflict
from mapper import TopicToFieldsMapper
from settings import get_settings


class UserData(BaseModel):
    influxdb_client: InfluxDBClient
    topic_to_fields: TopicToFieldsMapper
    data_aggregator: dict[str, InfluxDBLine] = Field(default_factory=dict)
    model_config = ConfigDict(arbitrary_types_allowed=True)


def on_connect(client: MQTTClient, userdata: UserData, flags, result_code: int) -> None:
    """The callback for when the client receives a CONNACK response from the server.

    Subscribing in on_connect() means that if we lose the connection and
    reconnect then subscriptions will be renewed.
    """
    del userdata, flags
    logger.info(f"Connected with result code {result_code}.")
    client.subscribe(get_settings().mqtt_topic_subscribe)


def on_message(client: MQTTClient, userdata: UserData, message: MQTTMessage) -> None:
    """The callback for when a PUBLISH message is received from the server."""
    del client
    line = userdata.topic_to_fields.to_influxdb_line(message.topic, message.payload)
    merge_id = line.merge_id

    logger.debug("Single measurement received.", line=line)

    try:
        userdata.data_aggregator[merge_id] = userdata.data_aggregator[merge_id].merge(line)
    except KeyError:
        userdata.data_aggregator[merge_id] = line
    except MergeConflict:
        write_to_influxdb(userdata.influxdb_client, userdata.data_aggregator.pop(merge_id))
        userdata.data_aggregator[merge_id] = line


def write_to_influxdb(client: InfluxDBClient, data: InfluxDBLine) -> None:
    with client.write_api(write_options=ASYNCHRONOUS) as write_api:
        write_api.write(bucket=data.bucket, record=data.dict())
    logger.debug("Data sent to InfluxDB", data=data)


def main() -> None:
    settings = get_settings()

    logger.configure(handlers=[])  # removes all pre-existing handlers
    logger.add(sys.stdout, format=LOGGING_FORMAT, level=settings.log_level, backtrace=True)

    InfluxDBLine.merge_data_points_on(settings.mqtt_merge_data_points_on)

    user_data = UserData(
        topic_to_fields=TopicToFieldsMapper(
            settings.mqtt_topic_pattern, settings.influxdb_default_bucket, settings.influxdb_default_measurement
        ),
        influxdb_client=InfluxDBClient(
            url=settings.influxdb_url,
            token=settings.influxdb_api_token,
            org=settings.influxdb_organization_id,
        ),
    )

    mqtt_client = MQTTClient(client_id=None)
    mqtt_client.username_pw_set(username=settings.mqtt_username, password=settings.mqtt_password)
    mqtt_client.user_data_set(user_data)

    mqtt_client.on_connect = on_connect
    mqtt_client.on_message = on_message

    mqtt_client.connect(host=str(settings.mqtt_host), port=settings.mqtt_port)

    try:
        mqtt_client.loop_forever()
    except KeyboardInterrupt:
        logger.info("Exiting.")


if __name__ == "__main__":
    main()


src/utils.py
from __future__ import annotations

from typing import Any, cast

from pydantic import BaseModel
from pydantic._internal._model_construction import ModelMetaclass
from pydantic.fields import FieldInfo


class _FieldInfoProxy:
    """Allows access to ``pydantic.FieldInfo`` attributes plus a name of this field.

    Name is not part of ``pydantic.FieldInfo``, so this proxy provides it. All other
    access is forwarded to the ``pydantic.FieldInfo`` instance.
    """

    def __init__(self, name: str, model_field: FieldInfo):
        self._name = name
        self.field_info = model_field

    def __getattr__(self, key: str) -> Any:
        if key == "name":
            return self._name
        return getattr(self.field_info, key)

    def __repr__(self):
        return self.field_info.__repr__()


class _FieldAccessor:
    """Allows static access to ``pydantic.BaseModel`` fields.

    You can access field attributes via ``Model.fields_.<field>.<attribute>``. Note the
    difference between ``fields`` and ``fields_``. ``fields`` is already defined on
    ``pydantic.BaseModel`` but it is an instance property and doesn't work on classes.

    Example:
        >>> from pydantic import BaseModel
        >>> class MyModel(BaseModel, metaclass=FieldAccessMeta):
        >>>     created_date: str
        >>> print(MyModel.fields_.created_date.name)
        "created_date"
    """

    def __init__(self, fields: dict[str, FieldInfo]):
        self._fields = fields

    def __getattr__(self, item: str) -> _FieldInfoProxy:
        try:
            return _FieldInfoProxy(item, self._fields[item])
        except KeyError as ex:
            raise AttributeError(f"{item} is not a model field.") from ex


class FieldAccessMeta(ModelMetaclass):
    """Adds static access to ``Pydantic.BaseModel`` fields and nested alias paths.

    See ``_FieldAccessor`` for details.
    """

    @property
    def fields_(cls) -> _FieldAccessor:
        """Static version of ``pydantic.BaseModel.fields``.

        The pydantic implementation is an instance property and doesn't work on classes.
        See ``_FieldAccessor`` for details.
        """
        base_model_cls = cast(BaseModel, cls)
        return _FieldAccessor(fields=base_model_cls.model_fields)


src/mapper.py
from copy import deepcopy
from typing import Optional, Union

from influx_db import InfluxDBLine, Tag


class TopicToFieldsMapper:
    _SPLIT_CHR = "/"
    _REQUIRED_FIELDS = {
        InfluxDBLine.fields_.bucket.name,
        InfluxDBLine.fields_.measurement.name,
        "field",
    }
    _AVAILABLE_FIELDS = deepcopy(_REQUIRED_FIELDS).union({InfluxDBLine.fields_.tags.name, "value_type"})

    def __init__(self, topic_pattern: str, default_bucket: Optional[str], default_measurement: Optional[str]):
        self._default_kwargs: dict[str, str] = {}
        if default_bucket:
            self._default_kwargs[InfluxDBLine.fields_.bucket.name] = default_bucket
        if default_measurement:
            self._default_kwargs[InfluxDBLine.fields_.measurement.name] = default_measurement
        self._pattern_parts = self._parse_patten(topic_pattern)

    def _parse_patten(self, topic_pattern: str) -> list[Union[str, Tag, None]]:
        parsed_variables: list[Union[str, Tag, None]] = []
        seen_variables: set[str] = set()

        for part in topic_pattern.split(self._SPLIT_CHR):
            if not part or part[0] != "{" or part[-1] != "}":
                parsed_variables.append(None)
                continue  # skip non-patterns
            part = part[1:-1]  # strip brackets
            if part.startswith("tag:"):
                parsed_variables.append(Tag(name=part[4:]))
                seen_variables.add(InfluxDBLine.fields_.tags.name)
            else:
                parsed_variables.append(part)
                seen_variables.add(part)

        if missing_vars := self._REQUIRED_FIELDS.difference(seen_variables.union(self._default_kwargs.keys())):
            raise ValueError(
                f"`mqtt_topic_pattern` set to '{topic_pattern}' is missing one or more "
                f"required variables: {missing_vars}. Make sure they are defined either "
                f"as part of the pattern or as the `influxdb_default_*` settings (if applicable)."
            )

        if unknown_vars := set(seen_variables).difference(self._AVAILABLE_FIELDS):
            raise ValueError(
                f"`mqtt_topic_pattern` set to '{topic_pattern}' contains unknown variable(s): {unknown_vars}"
            )

        return parsed_variables

    def to_influxdb_line(self, topic: str, value: str) -> InfluxDBLine:
        kwargs: dict[str, str] = deepcopy(self._default_kwargs)
        tags: dict[str, str] = {}

        for pattern, part in zip(self._pattern_parts, topic.split(self._SPLIT_CHR)):
            if not pattern:
                continue
            if isinstance(pattern, Tag):
                tags[pattern.name] = part
            else:
                kwargs[pattern] = part

        return InfluxDBLine.from_mqtt(value=value, tags=tags, **kwargs)


src/influx_db.py
from datetime import datetime, timezone
from typing import Any, ClassVar, Union

from pydantic import BaseModel, Field

from utils import FieldAccessMeta


class Tag(BaseModel):
    name: str


_Scalar = Union[str, int, bool, float]


class MergeConflict(Exception):
    pass


class InfluxDBLine(BaseModel, metaclass=FieldAccessMeta):
    measurement: str
    bucket: str
    fields: dict[str, Any]
    tags: dict[str, str] = Field(default_factory=dict)
    time: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    _VALID_TYPES: set[type] = {str, int, bool, float}
    _VALID_TYPE_NAMES: dict[str, type] = {_.__name__: _ for _ in _VALID_TYPES}

    _merge_on: ClassVar[str] = ""

    @classmethod
    def _parse_value(cls, value_type: str, value: str) -> _Scalar:
        if value_type not in cls._VALID_TYPE_NAMES:
            raise ValueError(
                f"'{value_type}' is not a valid value for 'value_type'. "
                f"Allowed values are: {list(cls._VALID_TYPE_NAMES.keys())}."
            )
        return cls._VALID_TYPE_NAMES[value_type](value)

    @classmethod
    def from_mqtt(
        cls,
        measurement: str,
        bucket: str,
        tags: dict[str, str],
        field: str,
        value: str,
        value_type: str = "str",
    ) -> "InfluxDBLine":
        return InfluxDBLine(
            measurement=measurement,
            bucket=bucket,
            fields={field: cls._parse_value(value_type, value)},
            tags=tags,
        )

    @classmethod
    def merge_data_points_on(cls, new_pattern: str) -> None:
        """Merge instances with identical string formatted using ``self.__dict__.`."""
        cls._merge_on = new_pattern

    @property
    def merge_id(self) -> str:
        return self._merge_on.format(**self.__dict__)

    def merge(self, other: "InfluxDBLine") -> "InfluxDBLine":
        if not isinstance(other, InfluxDBLine):
            raise TypeError(f"Can only merge instances of '{self.__class__.__name__}' but '{other.__class__}' used.")
        if self.merge_id != other.merge_id:
            raise ValueError("Only instances with identical `merge_id` can be merged.")
        if set(self.fields.keys()).intersection(other.fields.keys()):
            raise MergeConflict("Instances have overlapping fields.")

        model_copy = self.copy(deep=True)
        model_copy.fields.update(other.fields)

        return model_copy


tasks/docker_build.py
from getpass import getpass
from os import getenv
from pathlib import Path
from typing import List

from termcolor import cprint

try:
    import toml
except ImportError:
    print("This script depends on `toml`. Run `poetry add -D toml`.")
    raise

try:
    from packaging.version import Version
except ImportError:
    print("This script depends on `packaging`. Run `poetry add -D packaging`.")
    raise

from invoke import Context, task

from tasks.utils import print_header


def install_emulators(ctx: Context, build_for_platforms: List[str]) -> None:
    """See https://github.com/tonistiigi/binfmt#installing-emulators."""
    emulators = []
    if "linux/arm64" in build_for_platforms:
        emulators.append("arm64")
    if "linux/arm/v7" in build_for_platforms:
        emulators.append("arm")

    if emulators:
        print_header("Installing emulators", level=2, icon="⬇")
        ctx.run("docker run --privileged --rm tonistiigi/binfmt --install " + ",".join(emulators))


@task()
def docker_build(ctx, push=True):
    """Builds a docker image locally and pushes it to the image registry.

    Args:
        ctx (invoke.Context): Invoke context.
        push (bool):
            Push the built image and cache to remote docker image registry. Login is required
            and Personal Access Token will be requested from STDIN if `DOCKERHUB_PERSONAL_ACCESS_TOKEN`
            environment variable is not set.
    """
    print_header("RUNNING DOCKER BUILD")

    flags: List[str] = []

    pyproject = toml.load(Path(__file__).parent.parent / "pyproject.toml")
    project = pyproject["tool"]["poetry"]
    project_name = project["name"]
    project_version = project["version"]

    python_version = Version(project["dependencies"]["python"].strip("<>=~^"))
    flags.append(f"--build-arg PYTHON_VERSION='{python_version.major}.{python_version.minor}'")

    dockerhub = pyproject["tool"]["invoke"]["dockerhub"]
    dockerhub_username = dockerhub["username"]
    build_for_platforms = dockerhub["build_for_platforms"]
    flags.append("--platform " + ",".join(build_for_platforms))

    if push:
        dockerhub_password = getenv("DOCKERHUB_PERSONAL_ACCESS_TOKEN")
        if not dockerhub_password:
            cprint("DOCKERHUB_PERSONAL_ACCESS_TOKEN environment variable not set.", color="yellow")
            dockerhub_password = getpass("Dockerhub Personal Access Token: ")
        ctx.run(f"docker login --username {dockerhub_username} --password {dockerhub_password}")

        flags.append("--output type=image,push=true")

    if getenv("CI"):  # https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables
        flags.append("--progress plain")

    install_emulators(ctx, build_for_platforms)

    # While `docker buildx build` supports multiple `--tag` flags, push of them fails to expose
    # all architectures in `latest`. Multiple pushes fix this.
    for tag in [project_version, "latest"]:
        print_header(f"Build {dockerhub_username}/{project_name}:{tag}", level=2, icon="🔨")

        flags_for_tag = list(flags)
        if tag != "latest" and push:
            flags_for_tag.append(f"--cache-to type=registry,ref={dockerhub_username}/{project_name}")

        ctx.run(
            f"""
                docker buildx build \
                --cache-from type=registry,ref={dockerhub_username}/{project_name} \
                --tag {dockerhub_username}/{project_name}:{tag} \
                {" ".join(flags_for_tag)} \
                .
            """,
            echo=True,
            pty=True,
        )

    if push:
        ctx.run("docker logout")


src/settings.py
from functools import cache
from typing import Literal, Optional

from pydantic import Field, field_validator
from pydantic_settings import BaseSettings

from constants import LEVEL_DEFAULT, LOG_LEVELS


class Settings(BaseSettings):
    log_level: Literal["TRACE", "DEBUG", "INFO", "SUCCESS", "WARNING", "ERROR", "CRITICAL"] = Field(
        default=LEVEL_DEFAULT, description="Logging level."
    )

    # Configuration of access to InfluxDB Cloud.
    influxdb_url: str = Field(
        ...,
        description="URL of your InfluxDB Cloud instance.",
        pattern=r"https://.*\.influxdata.com",
    )
    influxdb_organization_id: str = Field(
        ...,
        description="Organization ID. A hexadecimal string. Should be part of the cloud "
        "instance URL or in About Organization. This is not the organization name!",
    )
    influxdb_api_token: str = Field(..., description="API token with write access to `influxdb_default_bucket`.")
    influxdb_precision: Literal["s", "ms", "ns", "us"] = Field(
        default="ns",
        description="The precision for the unix timestamps within the body line-protocol.",
    )
    influxdb_default_bucket: Optional[str] = Field(
        None,
        description="Default value if `{bucket}` in `mqtt_topic_pattern` below is not set or parsed.",
    )
    influxdb_default_measurement: Optional[str] = Field(
        None,
        description="Default value if `{measurement}` in `mqtt_topic_pattern` below is not set or parsed.",
    )

    mqtt_host: str = Field(..., description="IP address or a host name of an MQTT broker.")
    mqtt_port: int = Field(1883, min=1, max=65535, description="Port of the MQTT broker.")
    mqtt_username: Optional[str] = Field(None, description="User name used to authenticate with the MQTT broker.")
    mqtt_password: Optional[str] = Field(None, description="Password used to authenticate with the MQTT broker.")
    mqtt_topic_subscribe: str = Field(
        ...,
        description="A topic to subscribe to. You can also use the '+' and '#' wildcards.",
    )
    mqtt_topic_pattern: str = Field(
        ...,
        description="A pattern to parse the topic into fields sent to InfluxDB. "
        "Use `{variable}` or `{variable_type:value}` syntax in topic parts to create the "
        "match. Use a plain string to ignore a topic part. The pattern must be a prefix of "
        "`mqtt_topic_subscribe`.\n"
        "\n"
        "When both a variable and a `influxdb_*` configuration option are defined, variable takes precedence.",
        json_schema_extra={
            "possible_values": [
                (
                    "{bucket}",
                    "InfluxDB bucket to send data to. Required unless `influxdb_default_bucket` is set.",
                ),
                (
                    "{measurement}",
                    "InfluxDB measurement field. Required unless `influxdb_default_measurement` is set.",
                ),
                ("{field}", "Field name. Required."),
                (
                    "{value_type}",
                    "Value type used to parse the received data. Optional.\n  Acceptable values in the topic are "
                    "`str`, `int`, `float` and `bool`. Default value is `str`.",
                ),
                (
                    "{tag:TAG_NAME}",
                    "where TAG_NAME will be used as a tag name and the parsed value as a value. Optional.",
                ),
            ],
            "examples": (
                "`dt/influxdb/{bucket}/{measurement}/{tag:device_id}/{field}/{value_type}` "
                "will parse `dt/influxdb/home/environment/esp8266/temperature/float 23.412` into:\n"
                "\n"
                "```json\n"
                "{\n"
                '   "bucket": "home",\n'
                '   "measurement": "environment",\n'
                '   "tags": {"device_id": "esp8266"},\n'
                '   "fields": {"temperature": 23.412}\n'
                "}\n"
                "```"
            ),
        },
    )
    mqtt_merge_data_points_on: str = Field(
        default="{measurement}{bucket}{tags}",
        description="It is expected there will be a single value in each topic. But "
        "InfluxDB allows multiple fields to be sent at once (for example from one device). "
        "Use this option to define criteria on which to merge data points. The order and "
        "any extra characters aside from variables below are irrelevant. All collected fields "
        "are then sent to InfluxDB when any of the fields is seen more than once (suggesting "
        "new data is coming in).",
        json_schema_extra={
            "possible_values": [
                ("{bucket}",),
                ("{measurement}",),
                ("{tags}", "for all tags and their values"),
                ("{tags[NAME]}", "for a value of single tag named `NAME`"),
            ],
            "examples": (
                "A value of `{tags[device_id]}` will merge all received data that share the same `device_id` tag."
            ),
        },
    )

    # TODO[pydantic]: We couldn't refactor the `validator`, please replace it by `field_validator` manually.
    # Check https://docs.pydantic.dev/dev-v2/migration/#changes-to-validators for more information.
    @field_validator("log_level")
    def valid_log_level(cls, value, field):  # pylint: disable=no-self-argument
        if value not in LOG_LEVELS:
            raise ValueError(f"'{value}' is not allowed as a value for '{field}'. Allowed values are: {LOG_LEVELS}.")
        return value


@cache
def get_settings() -> Settings:
    return Settings()  # type: ignore[call-arg]
